### ISSUE TITLE
Can't load env config if something prevents global config file from being read

### ISSUE THREAD

**Issue by WhitWaldo** [cloudflare/wrangler#794]
### 🐛 Bug Report
I'm attempting to use wrangler within the Azure DevOps environment on a private build server using the environment variable approach to providing the configuration. I have confirmed that the env vars are being set during the build process, but when I attempt to use `wrangler publish`, I receive an error:
> Error: Your global config has an error, run `wrangler config`: missing field `email`.

It doesn't particularly matter if I use `wrangler config` and set up the global config beforehand, because DevOps will restrict access to the directory it's located within (e.g. can't really access files that don't live in the immediate build directory). I've verified this myself with a script that attempted to read the value of the config file (it existed) and I received an access denied error.

In reading how I'm hitting this error, I suspect it to be [this logic](https://github.com/cloudflare/wrangler/blob/80c9bfce38e67e671008a5452e90194584c0b540/src/settings/global_user.rs#L44) that first checks to see if the path exists, then attempts to read in the file. I expect that as in my own testing, this registers as truthy indicating that the path to the global config does indeed exist, but when you attempt to read it, you get an access denied. Because it's not otherwise wrapped with an exception handler, it bails at this point since what it fails to read is a valid config out of that file. 

I would suggest this be fixed by instead checking for the existence of the environment variables first, and then falling back to the presence of the config file if they are not populated. It would require only that lines 52-54 be inserted above line 44. It doesn't ultimately help much with providing a more useful error message if the file fails to be read, but would at least allow me to run wrangler with only environment variables configured sooner.

#### Environment

* operating system: Windows 2019 v1809
* output of `rustc -V`: Not installed
* output of `node -v`: v10.16.3
* output of `wrangler -V`: wrangler 1.4.0

#### Steps to reproduce

1. Since you use Azure DevOps, I'd suggest spinning up a private build server (since you'll only be able to repro this scenario if the config does exist, but can't be read). Pre-install nodejs and wrangler (globally) to it. Configure "wrangler config" and confirm that the configuration is written to the machine.

2. Generate a static site and save to a repo somewhere, create a new DevOps pipeline.

3. Set the environment variables to contain the equivalent to what you set up the configuration with and attempt to execute a "wrangler publish" operation. 

4. Observe the build failing with the error:
> Error: Your global config has an error, run `wrangler config`: missing field `email`

#### What did you expect to see?
Expected to see the environment variables used and the publish operation successfully completed.

#### What did you see instead?
Received the following error: 
> Error: Your global config has an error, run `wrangler config`: missing field `email`

**Comment by avidal** [comment]
Does it still fail if you *don't* have a global config? If wrangler config never runs there's never an unreadable but otherwise existing global config so the code path you hit shouldn't be encountered? Alternatively, what if you set WRANGLER_HOME=$(pwd) during the build process (or maybe something like `export WRANGLER_HOME=$(pwd)/fake-wrangler-config`, which gives you a directory with nothing in it, but it's within the build dir?

**Comment by ashleygwilliams** [comment]
hi @WhitWaldo ! long time no talk ;)

thanks for filing this with such good repro instructions. the detail is very helpful. i'll take a look at this and get back to you asap. can i ask why your CI has the wrangler config file? i'm a bit confused about that! thanks! -a

**Comment by WhitWaldo** [comment]
@ashleygwilliams Hello again - that's a fine question!

Unfortunately, the answer is because then, like today, I cannot get environment variables to work either. I figured the reasoning lies in me simply misconfiguring AzureDevOps somehow (was previously setting the variables in the job), but today I've set the variables on the machine itself (at a user-level, at a machine-level and then finally as both), deleted the config file and I'm still oddly seeing the same error message. 

![image](https://user-images.githubusercontent.com/2238529/67316635-70895700-f4ce-11e9-8074-ad9f0a55ebda.png)


In my package.json, I've set up wrangler to run in the local package:

"scripts": {
  "publish": "wrangler publish",
  "wrangler": "wrangler"
}

So in DevOps, I just call it with "npm run publish" and see:
![image](https://user-images.githubusercontent.com/2238529/67316568-4df73e00-f4ce-11e9-85e8-1a226109dc27.png)

This particular problem might be better suited to another issue though given that it stands separate from not being able to read an existing config, given that it's about not being able to get environment variables to work.

**Comment by EverlastingBugstopper** [comment]
Hmm.. this is strange. @WhitWaldo could you try setting your publish script to `CF_EMAIL=test@example.com wrangler publish` and see if it still gives you the same error message? It still wouldn't work of course but if you get a different error message it would let us see if the issue lies with global config or if it lies with Wrangler not seeing the environment variables you have set.

**Comment by WhitWaldo** [comment]
That did yield a slightly different error (DevOps isn't big on the npm variable assignment approach, so variable is set elsewhere):

![image](https://user-images.githubusercontent.com/2238529/67322343-7f740780-f4d6-11e9-86b6-105843ab8bb8.png)

**Comment by EverlastingBugstopper** [comment]
Ah! So this makes me think that for some reason `wrangler` is not picking up on the system-level environment variables that you have set since it was able to take in the email environment variable in this case. I'd make sure that running `echo $CF_EMAIL` and `echo $CF_API_KEY` print the values you expect. Environment variables should override everything in the configuration file, even if it exists and is completely empty.

Let us know if you have any luck!

**Comment by WhitWaldo** [comment]
Ok, I think you're on the right track - DevOps might be playing games with actual env variables since they do their own marshalling back and forth of them. I dropped in a step to verify that I was reading the values correctly and the CF_API_KEY wasn't being read properly. That's been resolved at this point, but unfortunately, wrangler still isn't a happy camper.

The debug text is pulling the appropriate values:
![image](https://user-images.githubusercontent.com/2238529/67336677-5f9b0e80-f4eb-11e9-9e23-8620e9ef12c6.png)

Unfortunately, we get this again in the npm step running wrangler publish:
![image](https://user-images.githubusercontent.com/2238529/67336745-7b061980-f4eb-11e9-8464-66a6933d81cc.png)

The variables I've got in place are "CF_EMAIL" and "CF_API_KEY"

Here's what the debug log of the task shows (though it's not much help, I'm afraid):

```
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'publish' ]
2 info using npm@6.9.0
3 info using node@v10.16.3
4 verbose run-script [ 'prepublish', 'publish', 'postpublish' ]
5 info lifecycle scripttestwebsite@1.0.0~prepublish: scripttestwebsite@1.0.0
6 info lifecycle scripttestwebsite@1.0.0~publish: scripttestwebsite@1.0.0
7 verbose lifecycle scripttestwebsite@1.0.0~publish: unsafe-perm in lifecycle true
8 verbose lifecycle scripttestwebsite@1.0.0~publish: PATH: <stuff in path>
9 verbose lifecycle scripttestwebsite@1.0.0~publish: CWD: C:\vstsagent\vsts-agent-win-x64-2.155.1\_work\33\s
10 silly lifecycle scripttestwebsite@1.0.0~publish: Args: [ '/d /s /c', 'wrangler publish' ]
11 silly lifecycle scripttestwebsite@1.0.0~publish: Returned: code: 1  signal: null
12 info lifecycle scripttestwebsite@1.0.0~publish: Failed to exec publish script
13 verbose stack Error: scripttestwebsite@1.0.0 publish: `wrangler publish`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:301:16)
13 verbose stack     at EventEmitter.emit (events.js:198:13)
13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:198:13)
13 verbose stack     at maybeClose (internal/child_process.js:982:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
14 verbose pkgid scripttestwebsite@1.0.0
15 verbose cwd C:\vstsagent\vsts-agent-win-x64-2.155.1\_work\33\s
16 verbose Windows_NT 10.0.17763
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "publish"
18 verbose node v10.16.3
19 verbose npm  v6.9.0
20 error code ELIFECYCLE
21 error errno 1
22 error scripttestwebsite@1.0.0 publish: `wrangler publish`
22 error Exit status 1
23 error Failed at the scripttestwebsite@1.0.0 publish script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

**Comment by WhitWaldo** [comment]
Any further ideas I can try out on this one? @EverlastingBugstopper

**Comment by ashleymichal** [comment]
Hey @WhitWaldo !

So I tested your initial ask, and it is true: Wrangler will indeed bail early on a file permissions error.  However, it appears to be getting past that step in your case. If Wrangler was running into a file permissions error on the config file, it would throw earlier than the error you are seeing (`Your global config has an error...`), and if the config file existed with appropriate permissions but was simply empty (another path I tested), it would fall back to the environment variables.

Without assessing your CI playbook, I can't say for certain, but one thing pops out at me: it appears that you are running `wrangler publish` via npm, probably using a script defined in your package.json? I would guess it is being invoked as `npm run publish`, where `publish` is defined in your package.json as `wrangler publish`. I know that node permissions and environment variables can get a little dicey, especially on a Windows machine, so I would recommend trying to invoke Wrangler directly and see if that improves your situation. Otherwise, the only other conclusion I can come to is that the environment variables are not being set before this step executes.
